src/Microsoft.OData.Core/Uri/NodeToStringBuilder.cs

```diff
@@ -470,7 +471,245 @@ internal String TranslateNode(QueryNode node)
+        /// <returns>The translated string with $it binding.</returns>
+        private string BindNode(QueryNode node, ResourceRangeVariable filterClauseRangeVariable)
+        {
+            BinaryOperatorNode binaryNode = node as BinaryOperatorNode;
+            InNode inNode = node as InNode;
+            AnyNode anyNode = node as AnyNode;
+            SingleValueFunctionCallNode singleValueFunctionCallNode = node as SingleValueFunctionCallNode;
+            SingleValueOpenPropertyAccessNode singleValueOpenPropertyAccessNode = node as SingleValueOpenPropertyAccessNode;
+
+            if (binaryNode != null)
+            {
+                return BindBinaryOperatorNode(binaryNode, filterClauseRangeVariable);
+            }
+            else if (inNode != null)
+            {
+                return BindInNode(inNode, filterClauseRangeVariable);
+            }
+            else if (anyNode != null)
+            {
+                return BindAnyNode(anyNode, filterClauseRangeVariable);
+            }
+            else if (singleValueFunctionCallNode != null)
+            {
+                return BindSingleValueFunctionCallNode(singleValueFunctionCallNode, filterClauseRangeVariable);
+            }
+            else if (singleValueOpenPropertyAccessNode != null)
+            {
+                return BindSingleValueOpenPropertyAccess(singleValueOpenPropertyAccessNode, filterClauseRangeVariable);
+            }
+            else
+            {
+                return this.TranslateNode(node);
+            }
+        }
+
+        /// <summary>
+        /// Bind $it to the <see cref="BinaryOperatorNode"/> translated string.
+        /// </summary>
+        /// <param name="node">node to bind.</param>
+        /// <param name="filterClauseRangeVariable">The <see cref="FilterClause"/> range variable.</param>
+        /// <returns>The translated string with $it binding.</returns>
+        private string BindBinaryOperatorNode(BinaryOperatorNode node, ResourceRangeVariable filterClauseRangeVariable)
+        {
+            string translatedNode = this.TranslateNode(node);
+            ResourceRangeVariableReferenceNode leftRangeVariableNode = GetResourceRangeVariableReferenceNode(node.Left);
+            ResourceRangeVariableReferenceNode rightRangeVariableNode = GetResourceRangeVariableReferenceNode(node.Right);
+
+            BinaryOperatorKind binaryNodeKind = node.OperatorKind;
+
+            string binaryOperator = this.BinaryOperatorNodeToString(binaryNodeKind);
+            string[] binarySeparator = { binaryOperator };
+            string[] substrings = translatedNode.Trim().Split(binarySeparator, StringSplitOptions.RemoveEmptyEntries);
+            string leftBinary = substrings[0].Trim();
+            string rightBinary = substrings[1].Trim();
+
+            if (leftRangeVariableNode != null && IsDifferentSource(filterClauseRangeVariable, leftRangeVariableNode))

```

Comment:
>IsDifferentSource(filterClauseRangeVariable, leftRangeVariableNode) [](start = 49, length = 67)

So this logic is assuming that, if the range variable is a different source, the range variable must be $it, right? In which case, $this would work because it's the same type as the filterClauseRangeVariable so it will just fall through? How would this logic change when if we supported $root?

Minimally, we need to document these assumptions and add tests (if they don't already exist) for these different cases.

Reply:
Good catch!